src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+            writer.Write("fallbackSourceFileEncoding", options.FallbackSourceFileEncoding?.CodePage);
+
+            writer.WriteObjectEnd();
+
+            static void writeSubsystemVersion(JsonWriter writer, SubsystemVersion version)
+            {
+                writer.WriteKey("subsystemVersion");
+                writer.WriteObjectStart();
+                writer.Write("major", version.Major);
+                writer.Write("minor", version.Minor);
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteCompilationOptions(JsonWriter writer, CompilationOptions options)
+        {
+            writer.WriteObjectStart();
+            WriteCompilationOptionsCore(writer, options);
+            writer.WriteObjectEnd();
+        }
+
+        protected virtual void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            // CompilationOption values
+            writer.Write("outputKind", options.OutputKind);
+            writer.Write("moduleName", options.ModuleName);
+            writer.Write("scriptClassName", options.ScriptClassName);
+            writer.Write("mainTypeName", options.MainTypeName);
+            WriteByteArrayValue(writer, "cryptoPublicKey", options.CryptoPublicKey);
+            writer.Write("cryptoKeyFile", options.CryptoKeyFile);
+            writer.Write("delaySign", options.DelaySign);
+            writer.Write("publicSign", options.PublicSign);
+            writer.Write("checkOverflow", options.CheckOverflow);
+            writer.Write("platform", options.Platform);
+            writer.Write("optimizationLevel", options.OptimizationLevel);
+            writer.Write("generalDiagnosticOption", options.GeneralDiagnosticOption);
+            writer.Write("warningLevel", options.WarningLevel);
+            writer.Write("deterministic", options.Deterministic);
+            writer.Write("debugPlusMode", options.DebugPlusMode);
+            writer.Write("referencesSupersedeLowerVersions", options.ReferencesSupersedeLowerVersions);
+            writer.Write("reportSuppressedDiagnostics", options.ReportSuppressedDiagnostics);
+            writer.Write("nullableContextOptions", options.NullableContextOptions);
+
+            writer.WriteKey("specificDiagnosticOptions");
+            writer.WriteArrayStart();
+            foreach (var kvp in options.SpecificDiagnosticOptions)
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
+            }
+            writer.WriteArrayEnd();
+
+            if (options.Deterministic)
+            {
+                writer.Write("deterministic", true);
+                writer.WriteNull("localtime");
+            }
+            else
+            {
+                writer.Write("deterministic", false);
+                writer.Write("localtime", options.CurrentLocalTime.ToString(CultureInfo.InvariantCulture));
```

Comment:
interesting.  so if we ask for a deterministic key for a compilatino we haven't passed the .deterministic option to, we are certain to get something not reusable? (not protesting... just interesting :))

Reply:
Hmm, good point as it's possible to have a case where you have `/deteriministic-`, the same local time but get different assemblies because with `/deterministic-` we effectively `Guid.NewGuid()` the MVID. Need to add an extra field here called `stopBeingNaughty: Guid.NewGuid()`